### ISSUE TITLE
Conversion success, but detect.tflite returns weird results.

### ISSUE THREAD

**Issue by OswinGuai** [tensorflow/models#9413]
# Prerequisites

Please answer the following questions for yourself before submitting an issue.

- [Y] I am reporting the issue to the correct repository. (Model Garden official or research directory)

## 1. The entire URL of the file you are using

https://github.com/tensorflow/models/tree/master/research/object_detection

## 2. Describe the bug

I fine-tuned an SSD-Mobilenet V2 model by Tensorflow-2.3, while converting the model to tflite by TF-Nightly. This works as I got a detect.tflite file of 11MB size. But when I use this model in my Android project, it outputs weird results, with extremely low confidences over all the 10 objects. As a comparison, I used Tensorflow-1.15 and TF-Nightly to generate a tflite model of SSD-Mobilenet V1, the result turned good. 

## 3. Steps to reproduce

As the following description,
- https://github.com/tensorflow/models/blob/master/research/object_detection/g3doc/tf2_training_and_evaluation.md
- https://github.com/tensorflow/models/blob/master/research/object_detection/g3doc/running_on_mobile_tf2.md

## 4. Expected behavior

I expect a detailed flow on how to generate a well working tflite model of SSD-Mobilenet V2.

## 5. Additional context

The bad results of SSD-Mobilenet V2,
![image](https://user-images.githubusercontent.com/7878475/97156835-3a9e4300-17b2-11eb-9ad4-409c75ee6ba4.png)

The good results of SSD-Mobilenet V1, 
![image](https://user-images.githubusercontent.com/7878475/97157043-86e98300-17b2-11eb-916b-cb7ae427758b.png)

## 6. System information

- Linux Ubuntu 16.04, Android.
- Simulator: Nexus_S_API_30
- Python version: 3.7

**Comment by OswinGuai** [comment]
Duplicate of #9287

**Comment by huberemanuel** [comment]
@OswinGuai you are sure that is a duplicate? Here the Mobilenet SSD v2 doesn't work well after conversion and in #9287 the same model works well after conversion, but the ResNet doesn't work well. I think they are the same issue, but now this functionality doesn't work with Mobilenet SSD v2.
I can relate the same issue with Mobilenet SSD v2, I will make some tests with tf1 version.

**Comment by OswinGuai** [comment]
@huberemanuel Sorry, my mistake, they are not duplicate. Your mention is right. I tried several times among versions but only got errors or wrong predictions. May you good luck.

**Comment by srjoglekar246** [comment]
@OswinGuai Can you give some pointers to your inference code? It seems like there is some difference in what the TF1 model did, vs TF2.

**Comment by huberemanuel** [comment]
Guys, finally got some good results with the converted tflite mobilenet SSD model. Firstly, I was trying to make some tests with tf1, but I moved back to tf2 after the newer version of tf-nightly corrected the less than 1kb tflite model size. The model prediction time was also increased, previous I was facing an issue that the original model was really faster than the tflite model, which didn't make sense (again, the new version of tf-nightly).
Now speaking of qualitative results, else they maintained poor prediction results, while the original model was around 80% accurate, the tflite model was not even 10% accurate. The solution was to normalize the input image as follows:

```python
ori_img = tf.keras.preprocessing.image.load_img(image_path, target_size=input_shape)
input_data = np.array(np.expand_dims(ori_img, 0))

# Important line
input_data = input_data / 128 - 1
```

After that, my tflite quantized model is presenting 80% accuracy just as the original model.
Hopefully, this helps you @OswinGuai to correct your issue.

If it helps, this is my tflite converting procedure:

```python
model = tf.saved_model.load(saved_model_dir)
converter = tf.lite.TFLiteConverter.from_saved_model(saved_model_dir=saved_model_dir, signature_keys=['serving_default'])
converter.optimizations = [tf.lite.Optimize.DEFAULT]
converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS,tf.lite.OpsSet.TFLITE_BUILTINS]
converter.experimental_new_converter = True
converter.allow_custom_ops = True
tflite_model = converter.convert()
```

**Comment by huberemanuel** [comment]
@srjoglekar246 the input normalization is a needed operation? I didn't found it in the tflite [example](https://github.com/tensorflow/models/blob/master/research/object_detection/colab_tutorials/eager_few_shot_od_training_tflite.ipynb)

**Comment by OswinGuai** [comment]
> @OswinGuai Can you give some pointers to your inference code? It seems like there is some difference in what the TF1 model did, vs TF2.

@srjoglekar246 Excuse me for so late response. Here is what I tried for inferencing by TensorFlow 2.*.
![image](https://user-images.githubusercontent.com/7878475/100686880-2ef9f980-33ba-11eb-91b7-b163a3f2a735.png)

**Comment by OswinGuai** [comment]
@huberemanuel It is cool. I will try it as soon as possible.

**Comment by srjoglekar246** [comment]
@huberemanuel Yes, you need to do preprocessing as mentioned [here](https://github.com/tensorflow/models/blob/master/research/slim/preprocessing/inception_preprocessing.py#L258). Could you check if there is a difference between this & your method?

**Comment by huberemanuel** [comment]
@srjoglekar246 the script you mentioned does the preprocessing I mentioned, but I think not in a way I was mentioning, I will rephrase it.
When I run the inference on an image after training with a SavedModel format, I don't actually need to preprocess the input to get the results. I believe this preprocessing step is done by the network. However, when I do this inference with a tflite I need to do this preprocessing by hand (or calling a function that does it), but the model itself doesn't take care of this step, leading to bad results if you don't normalize your image.
I just found in [exportf_tflite_graph_tf2.py](https://github.com/tensorflow/models/blob/master/research/object_detection/export_tflite_graph_tf2.py) that this normalization is indeed required, as it states "image: a float32 tensor of shape[1, height, width, 3] containing the *normalized* input image.", but this step is not done on the colab [example script](https://github.com/tensorflow/models/blob/master/research/object_detection/colab_tutorials/eager_few_shot_od_training_tflite.ipynb), so I think this script should be updated with this normalization (I can help with this), but more importantly, this normalization should be highlighted on the documentation.

**Comment by srjoglekar246** [comment]
@huberemanuel Agreed, the Inception pre-processing is a detail many people tend to miss. I will probably get to the documentation in a while, but (if you are interested), feel free to send a PR to [our documentation](https://github.com/tensorflow/models/blob/master/research/object_detection/g3doc/running_on_mobile_tf2.md).

**Comment by huberemanuel** [comment]
Just correcting what I said early, the colab example is doing the preprocessing, my mistake. I will make a PR highlighting this needed procedure, thank you @srjoglekar246

**Comment by srjoglekar246** [comment]
@OswinGuai Could you post your (untrained) tflite model OR pipeline config for me to take a look? And I assume you are using the SSD Android example (after modifying the app code parameters such as input size, etc if required?)

**Comment by Ethylbenzol** [comment]
> 
> 
> @huberemanuel Yes, you need to do preprocessing as mentioned [here](https://github.com/tensorflow/models/blob/master/research/slim/preprocessing/inception_preprocessing.py#L258). Could you check if there is a difference between this & your method?


I may have a similar issue as this, weird results after a success conversion. For normalized input images, do you mean the preprocessing in eval, image or train? [this](https://github.com/tensorflow/models/blob/master/research/slim/preprocessing/inception_preprocessing.py#L258) which you mention has 3 different function named preprocessing, and I am not sure which function should I call

**Comment by judahkshitij** [comment]
Hello, I am facing the same issue. Was there any update or resolution to this issue. I posted below comment in another thread that talks about wrong results from tflite model(s).

Hello @lechwolowski, @srjoglekar246 & Others, I have been trying to run inference after converting few models from TF2 obj detection zoo to TFLite using the process described in [this guide](https://github.com/tensorflow/models/blob/master/research/object_detection/g3doc/running_on_mobile_tf2.md), but getting wrong results from the tflite model (I am trying basic tflite model without doing any quantization or any other optimization as a first step).

The models I have tried are:

- ssd_mobilenet_v2_320x320_coco17_tpu-8
- ssd_mobilenet_v2_fpnlite_320x320_coco17_tpu-8
- ssd_mobilenet_v1_fpn_640x640_coco17_tpu-8
- ssd_resnet101_v1_fpn_640x640_coco17_tpu-8

The inference code I have used is the same as posted by @lechwolowski (also tried a few variants of the inference code that I found in other threads in this repo, but nothing worked).

I see that @lechwolowski was able to get correct results from above models except those using resnet. But for me, none of the above models are giving correct results on coco 2017 validation set images (even after tflite gets generated from the "tflite friendly" saved model). Can any of you provide insights on how you made it work or what I could be doing wrong? Any help is greatly appreciated. Thanks.

**Comment by UcefMountacer** [comment]
@srjoglekar246 
Thank you. Normalization improved result a LOT.
I just added this line:
`frame = frame / 255.`
Which I only used with my yolo tiny model. This is indeed not included in the tutorials.
Anyway thanks for the point.
